File: sdk-java/src/test/java/io/superdurable/dex/integ/RpcTest.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Client;
import io.superdurable.dex.StopFlowOptions;
import io.superdurable.dex.StopType;
import io.superdurable.dex.exceptions.FlowNotActiveException;
import io.superdurable.dex.exceptions.RpcLockConflictException;
import io.superdurable.dex.exceptions.WorkerInvocationException;
import io.superdurable.dex.testing.DexDevTestEnvironment;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

@Tag("dex-dev")
public final class RpcTest {
    private static final RpcNoStateWorkflow NO_STATE_WORKFLOW = new RpcNoStateWorkflow();
    private static final RpcWorkflow WORKFLOW = new RpcWorkflow();

    @TempDir
    Path cacheDirectory;

    @Test
    void testLockingRpc() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                NO_STATE_WORKFLOW)) {
            final String flowId = "rpc-lock-" + UUID.randomUUID();
            environment.client().startFlow(NO_STATE_WORKFLOW, flowId, null);
            final RpcNoStateWorkflow stub = environment.client().newRpcStub(
                    RpcNoStateWorkflow.class,
                    flowId);
            final ExecutorService executor = Executors.newFixedThreadPool(10);
            try {
                final List<Future<Boolean>> futures = new ArrayList<Future<Boolean>>();
                for (int index = 0; index < 100; index++) {
                    futures.add(executor.submit(() -> {
                        try {
                            environment.client().invokeRPC(stub::increaseCounter);
                            return true;
                        } catch (RpcLockConflictException conflict) {
                            return false;
                        }
                    }));
                }
                int succeeded = 0;
                for (Future<Boolean> future : futures) {
                    if (future.get()) {
                        succeeded++;
                    }
                }
                assertTrue(succeeded > 0);
                assertEquals(succeeded, environment.client().invokeRPC(stub::getCounter));
            } finally {
                executor.shutdownNow();
            }
            environment.client().stopFlow(flowId);
        }
    }

    @Test
    void testRpcProcedureWithoutAttributeAccess() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-no-attributes");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            environment.client().invokeRPC(stub::publishWithoutAttributeAccess);
            assertEquals(2, environment.client().waitForFlow(
                    flowId,
                    Integer.class,
                    Duration.ofSeconds(30)));
            assertThrows(
                    FlowNotActiveException.class,
                    () -> environment.client().invokeRPC(stub::publishWithoutAttributeAccess));
        }
    }

    @Test
    void testRpcWorkflowFunc1() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-func-1");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            environment.client().invokeRPC(stub::setData, "test-value");
            assertEquals("test-value", environment.client().invokeRPC(stub::getData));
            environment.client().invokeRPC(stub::setData, null);
            assertNull(environment.client().invokeRPC(stub::getData));
            environment.client().invokeRPC(stub::setKeyword, "test-value");
            assertEquals("test-value", environment.client().invokeRPC(stub::getKeyword));
            environment.client().invokeRPC(stub::setKeyword, null);
            assertNull(environment.client().invokeRPC(stub::getKeyword));
            assertEquals(
                    RpcWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::functionOne, "rpc-input"));
            assertRpcCompletion(environment, flowId, "rpc-input");
        }
    }

    @Test
    void testRpcWorkflowFunc0() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-func-0");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            assertEquals(
                    RpcWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::functionZero));
            assertRpcCompletion(environment, flowId, RpcWorkflow.HARDCODED_VALUE);
        }
    }

    @Test
    void testRpcWorkflowProc1() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-proc-1");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            environment.client().invokeRPC(stub::procedureOne, "rpc-input");
            assertRpcCompletion(environment, flowId, "rpc-input");
        }
    }

    @Test
    void testRpcWorkflowProc0() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-proc-0");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            environment.client().invokeRPC(stub::procedureZero);
            assertRpcCompletion(environment, flowId, RpcWorkflow.HARDCODED_VALUE);
        }
    }

    @Test
    void testRpcWorkflowFunc1ReadOnly() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                WORKFLOW)) {
            final String flowId = flowId("rpc-read-only");
            environment.client().startFlow(WORKFLOW, flowId, 999);
            final RpcWorkflow stub = stub(environment, flowId);
            assertEquals(
                    RpcWorkflow.RPC_OUTPUT,
                    environment.client().invokeRPC(stub::readOnly, "rpc-input"));
            environment.client().stopFlow(
                    flowId,
                    new StopFlowOptions(StopType.FAIL, RpcWorkflow.HARDCODED_VALUE));
        }
    }

    @Test
    void testRpcError() throws Exception {
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                NO_STATE_WORKFLOW)) {
            final String flowId = flowId("rpc-error");
            environment.client().startFlow(NO_STATE_WORKFLOW, flowId, null);
            final RpcNoStateWorkflow stub = environment.client().newRpcStub(
                    RpcNoStateWorkflow.class,
                    flowId);
            final WorkerInvocationException failure = assertThrows(
                    WorkerInvocationException.class,
                    () -> environment.client().invokeRPC(stub::fail, "this is an error"));
            assertTrue(failure.getWorkerErrorType().contains("IllegalArgumentException"));
            assertTrue(failure.getWorkerErrorDetail().contains("this is an error"));
            assertTrue(failure.getWorkerStackTrace().contains("IllegalArgumentException"));
            assertTrue(failure.getWorkerStackTrace().contains("this is an error"));
            environment.client().stopFlow(flowId);
        }
    }

    @Test
    void testSignalChannelSizeInfo() throws Exception {
        final NoStartStateDeadEndWorkflow workflow = new NoStartStateDeadEndWorkflow();
        try (DexDevTestEnvironment environment = DexDevTestEnvironment.start(
                cacheDirectory,
                workflow)) {
            final String flowId = flowId("channel-size");
            environment.client().startFlow(workflow, flowId, null);
            final NoStartStateDeadEndWorkflow stub = environment.client().newRpcStub(
                    NoStartStateDeadEndWorkflow.class,
                    flowId);
            environment.client().invokeRPC(stub::publishInternal);
            assertEquals(2, environment.client().invokeRPC(stub::publishInternal));
            environment.client().publish(
                    flowId,
                    workflow.idleSignal,
                    (Void) null,
                    (Void) null,
                    (Void) null);
            assertEquals(3, environment.client().invokeRPC(stub::signalSize));
            environment.client().stopFlow(flowId);
        }
    }

    void compileLocking(final Client client) {
        client.startFlow(NO_STATE_WORKFLOW, "rpc-lock", null);
        final RpcNoStateWorkflow stub = client.newRpcStub(
                RpcNoStateWorkflow.class,
                "rpc-lock");
        final Integer first = client.invokeRPC(stub::increaseCounter);
        final Integer second = client.invokeRPC(stub::getCounter);
        consume(first, second);
    }

    void compileFunctionsAndProcedures(final Client client) {
        client.startFlow(WORKFLOW, "rpc", 0);
        final RpcWorkflow stub = client.newRpcStub(
                RpcWorkflow.class,
                "rpc");
        client.invokeRPC(stub::publishWithoutAttributeAccess);
        final Long one = client.invokeRPC(stub::functionOne, "input");
        final Long zero = client.invokeRPC(stub::functionZero);
        client.invokeRPC(stub::procedureOne, "input");
        client.invokeRPC(stub::procedureZero);
        final Long readOnly = client.invokeRPC(stub::readOnly, "input");
        client.invokeRPC(stub::setData, "value");
        final String data = client.invokeRPC(stub::getData);
        client.invokeRPC(stub::setKeyword, "value");
        final String keyword = client.invokeRPC(stub::getKeyword);
        consume(one, zero, readOnly, data, keyword);
    }

    void compileRpcErrorAndChannelSize(final Client client) {
        final RpcNoStateWorkflow errorStub = client.newRpcStub(
                RpcNoStateWorkflow.class,
                "rpc-error");
        final Long ignored = client.invokeRPC(errorStub::fail, "error");
        final NoStartStateDeadEndWorkflow channelStub = client.newRpcStub(
                NoStartStateDeadEndWorkflow.class,
                "channel-size");
        final Integer published = client.invokeRPC(channelStub::publishInternal);
        final Integer size = client.invokeRPC(channelStub::signalSize);
        consume(ignored, published, size);
    }

    private static void consume(final Object... values) {
    }

    private static String flowId(final String prefix) {
        return prefix + "-" + UUID.randomUUID();
    }

    private static RpcWorkflow stub(
            final DexDevTestEnvironment environment,
            final String flowId) {
        return environment.client().newRpcStub(RpcWorkflow.class, flowId);
    }

    private static void assertRpcCompletion(
            final DexDevTestEnvironment environment,
            final String flowId,
            final String expectedValue) {
        assertEquals(2, environment.client().waitForFlow(
                flowId,
                Integer.class,
                Duration.ofSeconds(30)));
        assertEquals(expectedValue, environment.client().getAttribute(flowId, WORKFLOW.data));
        assertEquals(expectedValue, environment.client().getAttribute(flowId, WORKFLOW.keyword));
        assertEquals(
                Math.toIntExact(RpcWorkflow.RPC_OUTPUT),
                environment.client().getAttribute(flowId, WORKFLOW.integer));
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/rpc_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::sync::Arc;
use std::time::Duration;

use dex_sdk::{Client, GrpcCode, Registry, SdkError, SdkResult, StopFlowOptions};

use crate::no_start_state_dead_end_workflow::NoStartStateDeadEndWorkflow;
use crate::rpc_no_state_workflow::RpcNoStateWorkflow;
use crate::rpc_workflow::RpcWorkflow;
use crate::support::{DexDevTestEnvironment, flow_id};

#[test]
#[ignore = "requires dexcli dev"]
fn test_locking_rpc() {
    let environment = Arc::new(DexDevTestEnvironment::start(
        Registry::new().register(RpcNoStateWorkflow::new()),
    ));
    let workflow = RpcNoStateWorkflow::new();
    let flow_id = flow_id("rpc-lock");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start RPC locking Flow");

    let next_invocation = Arc::new(std::sync::atomic::AtomicUsize::new(0));
    let mut workers = Vec::new();
    for _ in 0..10 {
        let environment = Arc::clone(&environment);
        let flow_id = flow_id.clone();
        let next_invocation = Arc::clone(&next_invocation);
        workers.push(std::thread::spawn(move || {
            let mut succeeded = 0;
            loop {
                let invocation = next_invocation.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
                if invocation >= 100 {
                    return succeeded;
                }
                match environment
                    .client
                    .invoke_rpc_without_input::<i32>(&flow_id, RpcNoStateWorkflow::INCREASE_COUNTER)
                {
                    Ok(_) => succeeded += 1,
                    Err(SdkError::RpcLockConflict { .. }) => {}
                    Err(error) => panic!("increase counter RPC failed: {error:?}"),
                }
            }
        }));
    }
    let succeeded: usize = workers
        .into_iter()
        .map(|worker| worker.join().expect("join RPC worker"))
        .sum();
    assert!(succeeded > 0);
    assert_eq!(
        Some(succeeded as i32),
        environment
            .client
            .invoke_rpc_without_input::<Option<i32>>(&flow_id, RpcNoStateWorkflow::GET_COUNTER,)
            .expect("read locked counter")
    );
    environment
        .client
        .stop_flow(&flow_id, StopFlowOptions::cancel())
        .expect("stop RPC locking Flow");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_procedure_without_attribute_access() {
    let (environment, _workflow, flow_id) = start_rpc_workflow("rpc-no-attributes");
    environment
        .client
        .invoke_rpc_without_input(&flow_id, RpcWorkflow::PUBLISH_WITHOUT_ATTRIBUTE_ACCESS)
        .expect("publish without Attribute access");
    assert_eq!(
        2,
        environment
            .client
            .wait_for_flow_with_timeout::<i32>(&flow_id, Duration::from_secs(30))
            .expect("complete RPC Flow")
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_workflow_func1() {
    let (environment, workflow, flow_id) = start_rpc_workflow("rpc-func-1");
    verify_optional_rpc_attributes(&environment, &flow_id);
    assert_eq!(
        RpcWorkflow::RPC_OUTPUT,
        environment
            .client
            .invoke_rpc(&flow_id, RpcWorkflow::FUNCTION_ONE, "rpc-input".to_string())
            .expect("invoke function one")
    );
    assert_rpc_completion(&environment, &workflow, &flow_id, "rpc-input");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_workflow_func0() {
    let (environment, workflow, flow_id) = start_rpc_workflow("rpc-func-0");
    assert_eq!(
        RpcWorkflow::RPC_OUTPUT,
        environment
            .client
            .invoke_rpc_without_input(&flow_id, RpcWorkflow::FUNCTION_ZERO)
            .expect("invoke function zero")
    );
    assert_rpc_completion(
        &environment,
        &workflow,
        &flow_id,
        RpcWorkflow::HARDCODED_VALUE,
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_workflow_proc1() {
    let (environment, workflow, flow_id) = start_rpc_workflow("rpc-proc-1");
    environment
        .client
        .invoke_rpc(
            &flow_id,
            RpcWorkflow::PROCEDURE_ONE,
            "rpc-input".to_string(),
        )
        .expect("invoke procedure one");
    assert_rpc_completion(&environment, &workflow, &flow_id, "rpc-input");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_workflow_proc0() {
    let (environment, workflow, flow_id) = start_rpc_workflow("rpc-proc-0");
    environment
        .client
        .invoke_rpc_without_input(&flow_id, RpcWorkflow::PROCEDURE_ZERO)
        .expect("invoke procedure zero");
    assert_rpc_completion(
        &environment,
        &workflow,
        &flow_id,
        RpcWorkflow::HARDCODED_VALUE,
    );
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_workflow_func1_read_only() {
    let (environment, _workflow, flow_id) = start_rpc_workflow("rpc-read-only");
    assert_eq!(
        RpcWorkflow::RPC_OUTPUT,
        environment
            .client
            .invoke_rpc(&flow_id, RpcWorkflow::READ_ONLY, "rpc-input".to_string())
            .expect("invoke read-only function")
    );
    environment
        .client
        .stop_flow(
            &flow_id,
            StopFlowOptions::fail().reason(RpcWorkflow::HARDCODED_VALUE),
        )
        .expect("stop read-only Flow");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_rpc_error() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(RpcNoStateWorkflow::new()));
    let workflow = RpcNoStateWorkflow::new();
    let flow_id = flow_id("rpc-error");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start RPC error Flow");
    match environment
        .client
        .invoke_rpc::<String, i64>(
            &flow_id,
            RpcNoStateWorkflow::FAIL,
            "this is an error".to_string(),
        )
        .expect_err("RPC must return user error")
    {
        SdkError::WorkerInvocation {
            code,
            worker_error_type,
            worker_error_detail,
            ..
        } => {
            assert_eq!(GrpcCode::FailedPrecondition, code);
            assert!(worker_error_type.contains("HandlerError"));
            assert!(worker_error_detail.contains("this is an error"));
        }
        error => panic!("expected WorkerInvocation, got {error:?}"),
    }
    environment
        .client
        .stop_flow(&flow_id, StopFlowOptions::cancel())
        .expect("stop RPC error Flow");
}

#[test]
#[ignore = "requires dexcli dev"]
fn test_signal_channel_size_info() {
    let environment =
        DexDevTestEnvironment::start(Registry::new().register(NoStartStateDeadEndWorkflow::new()));
    let workflow = NoStartStateDeadEndWorkflow::new();
    let flow_id = flow_id("channel-size");
    environment
        .client
        .start_flow(&workflow, &flow_id, ())
        .expect("start channel-size Flow");
    environment
        .client
        .invoke_rpc_without_input::<usize>(&flow_id, NoStartStateDeadEndWorkflow::PUBLISH_INTERNAL)
        .expect("publish first internal message");
    assert_eq!(
        2,
        environment
            .client
            .invoke_rpc_without_input::<usize>(
                &flow_id,
                NoStartStateDeadEndWorkflow::PUBLISH_INTERNAL,
            )
            .expect("publish second internal message")
    );
    environment
        .client
        .publish_many(&flow_id, &workflow.idle_signal, [(), (), ()])
        .expect("publish external messages");
    assert_eq!(
        3,
        environment
            .client
            .invoke_rpc_without_input::<usize>(&flow_id, NoStartStateDeadEndWorkflow::SIGNAL_SIZE,)
            .expect("read signal size")
    );
    environment
        .client
        .stop_flow(&flow_id, StopFlowOptions::cancel())
        .expect("stop channel-size Flow");
}

fn start_rpc_workflow(prefix: &str) -> (DexDevTestEnvironment, RpcWorkflow, String) {
    let environment = DexDevTestEnvironment::start(Registry::new().register(RpcWorkflow::new()));
    let workflow = RpcWorkflow::new();
    let flow_id = flow_id(prefix);
    environment
        .client
        .start_flow(&workflow, &flow_id, 999)
        .expect("start RPC Flow");
    (environment, workflow, flow_id)
}

pub(crate) fn verify_optional_rpc_attributes(environment: &DexDevTestEnvironment, flow_id: &str) {
    environment
        .client
        .invoke_rpc(
            flow_id,
            RpcWorkflow::SET_DATA,
            Some("test-value".to_string()),
        )
        .expect("set data Attribute");
    assert_eq!(
        Some("test-value".to_string()),
        environment
            .client
            .invoke_rpc_without_input(flow_id, RpcWorkflow::GET_DATA)
            .expect("get data Attribute")
    );
    environment
        .client
        .invoke_rpc(flow_id, RpcWorkflow::SET_DATA, None)
        .expect("delete data Attribute");
    assert_eq!(
        None,
        environment
            .client
            .invoke_rpc_without_input::<Option<String>>(flow_id, RpcWorkflow::GET_DATA)
            .expect("get deleted data Attribute")
    );
    environment
        .client
        .invoke_rpc(
            flow_id,
            RpcWorkflow::SET_KEYWORD,
            Some("test-value".to_string()),
        )
        .expect("set keyword Attribute");
    assert_eq!(
        Some("test-value".to_string()),
        environment
            .client
            .invoke_rpc_without_input(flow_id, RpcWorkflow::GET_KEYWORD)
            .expect("get keyword Attribute")
    );
    environment
        .client
        .invoke_rpc(flow_id, RpcWorkflow::SET_KEYWORD, None)
        .expect("delete keyword Attribute");
    assert_eq!(
        None,
        environment
            .client
            .invoke_rpc_without_input::<Option<String>>(flow_id, RpcWorkflow::GET_KEYWORD)
            .expect("get deleted keyword Attribute")
    );
}

pub(crate) fn assert_rpc_completion(
    environment: &DexDevTestEnvironment,
    workflow: &RpcWorkflow,
    flow_id: &str,
    expected_value: &str,
) {
    assert_eq!(
        2,
        environment
            .client
            .wait_for_flow_with_timeout::<i32>(flow_id, Duration::from_secs(30))
            .expect("complete RPC Flow")
    );
    assert_eq!(
        Some(expected_value.to_string()),
        environment
            .client
            .get_attribute(flow_id, &workflow.data)
            .expect("get data Attribute")
    );
    assert_eq!(
        Some(expected_value.to_string()),
        environment
            .client
            .get_attribute(flow_id, &workflow.keyword)
            .expect("get keyword Attribute")
    );
    assert_eq!(
        Some(RpcWorkflow::RPC_OUTPUT as i32),
        environment
            .client
            .get_attribute(flow_id, &workflow.integer)
            .expect("get integer Attribute")
    );
}

#[allow(dead_code)]
fn compile_locking(client: &Client) -> SdkResult<()> {
    client.start_flow(&RpcNoStateWorkflow::new(), "rpc-lock", ())?;
    let _first: i32 =
        client.invoke_rpc_without_input("rpc-lock", RpcNoStateWorkflow::INCREASE_COUNTER)?;
    let _second: Option<i32> =
        client.invoke_rpc_without_input("rpc-lock", RpcNoStateWorkflow::GET_COUNTER)?;
    Ok(())
}

#[allow(dead_code)]
fn compile_functions_and_procedures(client: &Client) -> SdkResult<()> {
    client.start_flow(&RpcWorkflow::new(), "rpc", 0)?;
    client.invoke_rpc_without_input::<()>("rpc", RpcWorkflow::PUBLISH_WITHOUT_ATTRIBUTE_ACCESS)?;
    let _one: i64 = client.invoke_rpc("rpc", RpcWorkflow::FUNCTION_ONE, "input".to_string())?;
    let _zero: i64 = client.invoke_rpc_without_input("rpc", RpcWorkflow::FUNCTION_ZERO)?;
    client.invoke_rpc("rpc", RpcWorkflow::PROCEDURE_ONE, "input".to_string())?;
    client.invoke_rpc_without_input::<()>("rpc", RpcWorkflow::PROCEDURE_ZERO)?;
    let _read_only: i64 = client.invoke_rpc("rpc", RpcWorkflow::READ_ONLY, "input".to_string())?;
    client.invoke_rpc("rpc", RpcWorkflow::SET_DATA, Some("value".to_string()))?;
    let _data: Option<String> = client.invoke_rpc_without_input("rpc", RpcWorkflow::GET_DATA)?;
    client.invoke_rpc("rpc", RpcWorkflow::SET_KEYWORD, Some("value".to_string()))?;
    let _keyword: Option<String> =
        client.invoke_rpc_without_input("rpc", RpcWorkflow::GET_KEYWORD)?;
    Ok(())
}

#[allow(dead_code)]
fn compile_rpc_error_and_channel_size(client: &Client) -> SdkResult<()> {
    let _ignored: i64 =
        client.invoke_rpc("rpc-error", RpcNoStateWorkflow::FAIL, "error".to_string())?;
    let _published: usize = client.invoke_rpc_without_input(
        "channel-size",
        NoStartStateDeadEndWorkflow::PUBLISH_INTERNAL,
    )?;
    let _size: usize = client
        .invoke_rpc_without_input("channel-size", NoStartStateDeadEndWorkflow::SIGNAL_SIZE)?;
    Ok(())
}
